torchtitan/experiments/kernels/triton_mg_group_gemm/benchmark.py

```diff
@@ -0,0 +1,619 @@
+#!/usr/bin/env python3
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# Benchmark comparing reference PyTorch vs optimized M*G group GEMM implementation
+
+import argparse
+import logging
+import time
+from typing import Dict, List, Optional, Tuple
+
+import matplotlib.pyplot as plt
+import numpy as np
+import torch
+import triton
+import triton.language as tl
+
+# Configure logging
+logging.basicConfig(
+    level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s"
+)
+
+# Try to import the optimized implementations
+try:
+    from mg_backward import grouped_gemm_backward
+    from mg_forward import grouped_gemm_forward
+except ImportError:
+    logging.error(
+        "Error importing MG grouped GEMM modules. Make sure the implementation files are in the correct path."
+    )
+    raise
+
+
+def compute_reference_forward(x, w, m_sizes):
+    """
+    Reference PyTorch implementation of M*G grouped GEMM forward pass.
+
+    Args:
+        x (torch.Tensor): Input tensor of shape (M, K)
+        w (torch.Tensor): Weight tensor of shape (N, K)
+        m_sizes (torch.Tensor): Group sizes tensor of shape (G)
+
+    Returns:
+        torch.Tensor: Output tensor of shape (M, N)
+    """
+    result = torch.zeros((x.shape[0], w.shape[0]), dtype=x.dtype, device=x.device)
+
+    m_start = 0
+    for g in range(len(m_sizes)):
+        m_size = m_sizes[g].item()
+        if m_size > 0:
+            m_end = m_start + m_size
+
+            # Extract group input
+            x_g = x[m_start:m_end]
+
+            # Compute group output
+            y_g = torch.matmul(x_g, w.T)
```

Comment:
Just want to make sure i understand correctly, `w` is similar to weight matrix of a *single* expert, rather than all experts glued together, right?

Reply:
yes currently this is correct.
That's actually the next problem - we have a couple options for how to further group this (concatenate, batch, etc).

Reply:
Thanks. 
In the benchmar, we call like this:
```
grouped_gemm_forward(x, w, m_sizes)  <-- individual weight
```
Whereas in `simple_MoE.py`, we call like this:
```
hidden_outputs = grouped_gemm_forward(
    combined_input, fc1_weight_combined, m_sizes  <-- grouped weight
)
```
So I was wondering what the definition is.

Reply:
Yes exactly.  I think this combined [MG, NE] approach is likely optimal for the implementation but we'll see in DeepSeek.